Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/ThresholdedRandomCutForest.java

```diff
@@ -240,6 +243,89 @@ public AnomalyDescriptor process(double[] inputPoint, long timestamp, int[] miss
+     * missing data, the internal imputation and extrapolation need not be
+     * consistent.
+     * 
+     * @param horizon    the length of time in the future which is being forecast
+     * @param correct    a boolean indicating if predictor-corrector subroutine
+     *                   should be turned on; this is specially helpful if there has
+     *                   been an anomaly in the recent past
+     * @param centrality in general RCF predicts the p50 value of conditional
+     *                   samples (centrality = 1). This parameter relaxes the
+     *                   conditional sampling. Using assumptions about input data
+     *                   (hence external to this code) it may be possible to use
+     *                   this parameter and the range information for confidence
+     *                   bounds.
+     * @return a timed range vector where the values[i] correspond to the forecast
+     *         for horizon (i+1). The upper and lower arrays indicate the
+     *         corresponding bounds based on the conditional sampling (and
+     *         transformation). Note that TRCF manages time in process() and thus
+     *         the forecasts always have timestamps associated which makes it easier
+     *         to execute the same code for various forest modes such as
+     *         STREAMING_IMPUTE, STANDARD and TIME_AUGMENTED. For STREAMING_IMPUTE
+     *         the time components of the prediction will be 0 because the time
+     *         information is already being used to fill in missing entries. For
+     *         STANDARD mode the time components would correspond to average arrival
+     *         difference. For TIME_AUGMENTED mode the time componentes would be the
+     *         result of the joint prediction. Finally note that setting weight of
+     *         time or any of the input columns will also 0 out the corresponding
+     *         forecast.
+     */
+
+    public TimedRangeVector extrapolate(int horizon, boolean correct, double centrality) {
+        // checkArgument(forestMode != ForestMode.STREAMING_IMPUTE, "not yet
+        // supported");
+        checkArgument(
+                (transformMethod != TransformMethod.DIFFERENCE
+                        && transformMethod != TransformMethod.NORMALIZE_DIFFERENCE)
+                        || horizon <= preprocessor.getShingleSize() / 2 + 1,
+                "reduce horizon or use a different transformation, single step differencing will be noisy");
+        int shingleSize = preprocessor.getShingleSize();
+        checkArgument(shingleSize > 1, "extrapolation is not meaningful for shingle size = 1");
+        // note the forest may have external shingling ...
+        int dimensions = forest.getDimensions();
+        int blockSize = dimensions / shingleSize;
+        double[] lastPoint = preprocessor.getLastShingledPoint();
+        boolean ifZero = (forest.getBoundingBoxCacheFraction() == 0);
```

Comment:
what if we called this `cacheDisabled`.

Reply:
Done. Also fixed at the earlier place in singleStepProcess().